bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/configurable/providers/toml/TomlProvider.java

```diff
@@ -512,74 +514,59 @@ private void addToModuleNodeMap(Toml table, List<TomlTableNode> moduleNodes) {
-        if (tableNode.isPresent()) {
-            errorNode = tableNode.get().rootNode();
-        } else if (valueNode.isPresent()) {
-            errorNode = valueNode.get();
-        } else if (!tomlTables.isEmpty()) {
-            errorNode = tomlTables.get(0).rootNode();
+        String orgModuleKey = getModuleKey(module);
+        Optional<TomlNode> optionalErrorNode = getErrorNode(toml, module.getName(), orgModuleKey, variableName);
+        if (optionalErrorNode.isPresent()) {
+            errorNode = optionalErrorNode.get();
         }
         if (errorNode != null) {
             invalidRequiredModuleSet.add(module.toString());
             invalidTomlLines.add(errorNode.location().lineRange());
-            throw new ConfigException(CONFIG_TOML_INVALID_MODULE_STRUCTURE, getLineRange(errorNode), moduleKey,
-                    moduleKey);
+            throw new ConfigException(CONFIG_TOML_INVALID_MODULE_STRUCTURE_WITH_VARIABLE, getLineRange(errorNode),
+                    orgModuleKey, variableName, orgModuleKey);
         }
     }
 
-    private List<TomlTableNode> getNonDefaultModuleNode(Toml baseToml, Module module, boolean hasRequired) {
+    private List<TomlTableNode> getNonDefaultModuleNode(Toml baseToml, Module module, boolean hasRequired,
+                                                        String variableName) {
         String moduleName = module.getName();
         Optional<Toml> table;
         List<TomlTableNode> moduleNodes = new ArrayList<>();
-        String moduleKey = getModuleKey(module);
+        String orgModuleKey = getModuleKey(module);
         if (moduleInfo.hasModuleAmbiguity()) {
-            table = baseToml.getTable(moduleKey);
+            table = baseToml.getTable(orgModuleKey);
             if (table.isPresent()) {
                 addToModuleNodeMap(table.get(), moduleNodes);
-            } else if (!invalidRequiredModuleSet.contains(module.toString()) && hasRequired) {
+            } else if (hasRequired) {
                 invalidRequiredModuleSet.add(module.toString());
                 throw new ConfigException(RuntimeErrors.CONFIG_TOML_MODULE_AMBIGUITY, getLineRange(baseToml.rootNode()),
-                        moduleName, moduleKey);
+                        moduleName, orgModuleKey);
             }
         }
         table = baseToml.getTable(moduleName);
-        table.ifPresent(toml -> moduleNodes.add(toml.rootNode()));
-        table = baseToml.getTable(moduleKey);
-        if (table.isEmpty() && hasRequired && !invalidRequiredModuleSet.contains(module.toString())) {
-            throwInvalidSubModuleError(baseToml, module);
+        boolean tableFound = table.isPresent();
+        if (tableFound) {
+            addToModuleNodeMap(table.get(), moduleNodes);
+        }
+        table = baseToml.getTable(orgModuleKey);
+        if (table.isPresent()) {
+            addToModuleNodeMap(table.get(), moduleNodes);
+            tableFound = true;
```

Comment:
Should tableFound is always true?

Reply:
Since table.isPresent() is assigned to tableFound last line

Reply:
As the `table` variable is assigned twice in the code for 
1. module name 
```java
        table = baseToml.getTable(moduleName);
```
2. module name with org name
```java 
        table = baseToml.getTable(orgModuleKey);
```

We need to reassign the boolean variable `tableFound` again and use that to throw the  error.

Reply:
As per the offline discussion, we can't remove `addToModuleNodeMap` at each toml retrieval. Because we need to collect all module representations (with module name/ module+org name) in the TOML file. I removed the `tableFound` boolean and modified the check to use `moduleNodes` list.